packages/graphiql/src/components/HeaderEditor.tsx

```diff
@@ -59,21 +61,68 @@ export class HeaderEditor extends React.Component<HeaderEditorProps> {
   }
 
   componentDidMount() {
-    // Lazily require to ensure requiring GraphiQL outside of a Browser context
-    // does not produce an error.
-    this.CodeMirror = require('codemirror');
-    require('codemirror/addon/hint/show-hint');
-    require('codemirror/addon/edit/matchbrackets');
-    require('codemirror/addon/edit/closebrackets');
-    require('codemirror/addon/fold/brace-fold');
-    require('codemirror/addon/fold/foldgutter');
-    require('codemirror/addon/lint/lint');
-    require('codemirror/addon/search/searchcursor');
-    require('codemirror/addon/search/jump-to-line');
-    require('codemirror/addon/dialog/dialog');
-    require('codemirror/mode/javascript/javascript');
-    require('codemirror/keymap/sublime');
+    this.initializeEditor()
+      .then(editor => {
+        editor.on('change', this._onEdit);
+        editor.on('keyup', this._onKeyUp);
+        editor.on('hasCompletion', this._onHasCompletion);
+      })
+      .catch(console.error);
+  }
+
+  componentDidUpdate(prevProps: HeaderEditorProps) {
+    if (!this.editor) {
+      return;
+    }
+
+    // Ensure the changes caused by this update are not interpretted as
+    // user-input changes which could otherwise result in an infinite
+    // event loop.
+    this.ignoreChangeEvent = true;
+    if (
+      this.props.value !== prevProps.value &&
+      this.props.value !== this.cachedValue
+    ) {
+      const thisValue = this.props.value || '';
+      this.cachedValue = thisValue;
+      this.editor.setValue(thisValue);
+    }
+    this.ignoreChangeEvent = false;
+  }
+
+  componentWillUnmount() {
+    if (!this.editor) {
+      return;
+    }
+    this.editor.off('change', this._onEdit);
+    this.editor.off('keyup', this._onKeyUp);
+    // @ts-expect-error
+    this.editor.off('hasCompletion', this._onHasCompletion);
+    this.editor = null;
+  }
+
+  render() {
+    return (
+      <div
+        className="codemirrorWrap"
+        // This horrible hack is necessary because a simple display none toggle
+        // causes one of the editors' gutters to break otherwise.
+        style={{
+          position: this.props.active ? 'relative' : 'absolute',
+          visibility: this.props.active ? 'visible' : 'hidden',
+        }}
+        ref={node => {
+          this._node = node as HTMLDivElement;
+        }}
+      />
+    );
+  }
+
+  // @ts-expect-error
+  addonModules = () => [import('codemirror/mode/javascript/javascript')];
```

Comment:
What would the TS error be here?

Reply:
that there isn't a type definition for this for `@types/codemirror`, just a legacy codemirror 5 issue

Reply:
Ok, thanks for the context! So the afterworld can find this context in this PR

Reply:
I could just properly override them in a .d.ts file I guess, but this was easier?